drivers/ethernet/eth_nxp_enet.c

```diff
@@ -0,0 +1,1045 @@
+
+				prio = net_vlan2priority(
+						net_pkt_vlan_priority(pkt));
+				net_pkt_set_priority(pkt, prio);
+			}
+#endif /* CONFIG_NET_TC_RX_COUNT > 1 */
+		}
+	}
+#endif /* CONFIG_NET_VLAN */
+
+	/*
+	 * Use MAC timestamp
+	 */
+#if defined(CONFIG_PTP_CLOCK_NXP_ENET)
+	k_mutex_lock(data->ptp_mutex, K_FOREVER);
+	if (eth_get_ptp_data(get_iface(data, vlan_tag), pkt)) {
+		ENET_Ptp1588GetTimer(config->base, &data->enet_handle,
+					&ptp_time_data);
+		/* If latest timestamp reloads after getting from Rx BD,
+		 * then second - 1 to make sure the actual Rx timestamp is
+		 * accurate
+		 */
+		if (ptp_time_data.nanosecond < ts) {
+			ptp_time_data.second--;
+		}
+
+		pkt->timestamp.nanosecond = ts;
+		pkt->timestamp.second = ptp_time_data.second;
+	} else {
+		/* Invalid value. */
+		pkt->timestamp.nanosecond = UINT32_MAX;
+		pkt->timestamp.second = UINT64_MAX;
+	}
+	k_mutex_unlock(data->ptp_mutex);
+#endif /* CONFIG_PTP_CLOCK_NXP_ENET */
+
+	iface = get_iface(data, vlan_tag);
+#if defined(CONFIG_NET_DSA)
+	iface = dsa_net_recv(iface, &pkt);
+#endif
+	if (net_recv_data(iface, pkt) < 0) {
+		net_pkt_unref(pkt);
+		goto error;
+	}
+
+	return 1;
+flush:
+	/* Flush the current read buffer.  This operation can
+	 * only report failure if there is no frame to flush,
+	 * which cannot happen in this context.
+	 */
+	status = ENET_ReadFrame(config->base, &data->enet_handle, NULL,
+					0, RING_ID, NULL);
+	__ASSERT_NO_MSG(status == kStatus_Success);
+error:
+	eth_stats_update_errors_rx(get_iface(data, vlan_tag));
+	return -EIO;
+}
+
+static void eth_nxp_enet_rx_thread(void *arg1, void *unused1, void *unused2)
+{
+	const struct device *dev = arg1;
+	const struct nxp_enet_mac_config *config = dev->config;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	while (1) {
+		if (k_sem_take(&data->rx_thread_sem, K_FOREVER) == 0) {
+			while (eth_nxp_enet_rx(dev) == 1) {
+				;
+			}
+			/* enable the IRQ for RX */
+			ENET_EnableInterrupts(config->base,
+			  kENET_RxFrameInterrupt | kENET_RxBufferInterrupt);
+		}
+	}
+}
+
+/*
+ ****************************
+ * PHY management functions *
+ ****************************
+ */
+
+static int nxp_enet_phy_reset_and_configure(const struct device *phy)
+{
+	int ret = 0;
+
+	/* Reset the PHY */
+	ret = phy_write(phy, MII_BMCR, MII_BMCR_RESET);
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure the PHY */
+	ret = phy_configure_link(phy, LINK_HALF_10BASE_T |
+				LINK_FULL_10BASE_T |
+				LINK_HALF_100BASE_T |
+				LINK_FULL_100BASE_T);
+	if (ret) {
+		return ret;
+	}
+
+	return ret;
+}
+
+static void nxp_enet_phy_cb(const struct device *phy,
+				struct phy_link_state *state,
+				void *eth_dev)
+{
+	struct phy_link_state new_state;
+	const struct device *dev = eth_dev;
+	struct nxp_enet_mac_data *data = dev->data;
+
+	if (!state->is_up) {
+		net_eth_carrier_off(data->iface);
+		/* If link went down, keep trying to re-establish it */
+		while (phy_get_link_state(phy, &new_state) || !new_state.is_up) {
```

Comment:
is this really necessary? `phy_mc_ksz8081_monitor_work_handler` is periodically checking for the link status. I'd suggest to instead bring the carrier down/up according to the link status passed to this cb.

Reply:
I am not sure what you mean with this, do you mean the phy driver should try to re-establish? or do you mean no reset is necessary and just wait more patiently?

Reply:
If I understand correctly, the phy driver has delayable task monitoring the status of the link every `CONFIG_PHY_MONITOR_PERIOD` ms and on each link state change it will execute `nxp_enet_phy_cb` 
https://github.com/zephyrproject-rtos/zephyr/pull/62833/files#diff-8264e34cf3eb87668d72eefdb8154c2a3d452eea4e666168e8f43e01d9795306R374

When the link goes down, this callback will start resetting and re-configuring the phy and polling for the link status, I believe at the same time as the phy monitor task continue running. My suggestion is to let only the phy driver to monitor the link status at a fixed rate, and inform the enet driver when there's a link state change to do the needed re-configurations and notifications of the carrier status to the networking subsys. Then on this callback:
```c
if (state->is_up) {
    // perform any reconfiguration needed
    net_eth_carrier_on(data->iface);
} else {
    net_eth_carrier_off(data->iface);
}
```
For example, https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_sam_gmac.c#L1831-L1860
Perhaps I'm missing some piece to understand why this polling is needed.

Reply:
oh, I understand what you mean now, yes I agree, this was a mistake, since this callback function is already periodic it doesnt make sense to nest more periods, I will change it